doc/cfg_reload_xact_enhance/Config_Reload_Enhancement_Xact_HLD.md

```diff
@@ -0,0 +1,230 @@
+
+### Requirements
+
+This enhancement requires the mechanism to know whether the device has reached an operational state to decide the finish of a config reload transaction.
+The system health feature provides the functionality to monitor the status of the sonic related services and the sonic critical processes. This enhancement will leverage the functionality to decide the current operational state. Refer to [SONiC System Health Monitor High Level Design](https://github.com/Azure/SONiC/blob/master/doc/system_health_monitoring/system-health-HLD.md) for the details of the system health feature.
+
+### Architecture Design 
+
+There is no change in the SONiC architecture and no new module/submodule is added.
+
+### High-Level Design 
+
+This enhancement is based on the existing 'config reload' Sonic CLI command located in 'sonic-utilities'. The modifications of this enhancement are in the file where the code for 'config reload' lies in.
+
+For the brevity of the explanation, two process names are defined here and will be used in the following paragraphs. The first process name 'CFG_RELOAD_PROC' is used to represent the process that runs the 'config reload' command. The second process name 'MONITOR_RELOAD_PROC' is used to represent the config reload monitor process, which is forked by 'CFG_RELOAD_PROC' and is daemonized to monitor the transitions of the overall config reload transaction.
+
+A high level sequential diagram is shown below.
+<img src="files/sequential-diagram.png" alt="sequential-diagram" width="1245px"/>
+
+In order to guarantee only one 'config reload' session is allowed to be run in the system, it is required to acquire a transaction lock before the operation starts. When the command 'config reload' is executed, it acquires the lock. If the lock is acquired successfully, it continues the operation. If the lock cannot be acquired, it checks the running transaction by the existence of 'CFG_RELOAD_PROC' and 'MONITOR_RELOAD_PROC' through the output of 'ps -ax'. It continues the operation when no running transaction is found, and it is not necessary to do the lock operation in this condition because the lock is already in locked state. If any running transaction is detected, it aborts the operation and outputs the error message immediately. The flow chart shown below illustrates the flow to start a config reload transaction.
+<img src="files/init-xact-flow-chart.png" alt="init-xact-flow-chart" width="1323px"/>
+
+The detection of running transaction enables the resilience. Assume that the transaction lock is not released correctly by 'MONITOR_RELOAD_PROC' due to an unexpected error upon the previous 'config reload', the error condition can be resolved automatically by the detection.
```

Comment:
Not sure that we need a lock here since you are checking the the process existence by `ps` command.

Reply:
I do have the same question, we could very well check the ps and determine whether to allow another 'config reload' or not.